tls/s2n_async_pkey.c

```diff
@@ -191,154 +197,154 @@ int s2n_async_pkey_sign_async(struct s2n_connection *conn, s2n_signature_algorit
-                             struct s2n_hash_state *digest, s2n_async_pkey_sign_complete on_complete)
+S2N_RESULT s2n_async_pkey_sign_sync(struct s2n_connection *conn, s2n_signature_algorithm sig_alg,
+                                    struct s2n_hash_state *digest, s2n_async_pkey_sign_complete on_complete)
 {
     const struct s2n_pkey *pkey = conn->handshake_params.our_chain_and_key->private_key;
     DEFER_CLEANUP(struct s2n_blob signed_content = { 0 }, s2n_free);
 
     uint32_t maximum_signature_length = s2n_pkey_size(pkey);
-    GUARD(s2n_alloc(&signed_content, maximum_signature_length));
+    GUARD_AS_RESULT(s2n_alloc(&signed_content, maximum_signature_length));
 
-    GUARD(s2n_pkey_sign(pkey, sig_alg, digest, &signed_content));
+    GUARD_AS_RESULT(s2n_pkey_sign(pkey, sig_alg, digest, &signed_content));
 
-    GUARD(on_complete(conn, &signed_content));
+    GUARD_AS_RESULT(on_complete(conn, &signed_content));
 
-    return 0;
+    return S2N_RESULT_OK;
 }
 
 int s2n_async_pkey_op_perform(struct s2n_async_pkey_op *op, s2n_cert_private_key *key)
 {
-    notnull_check(op);
-    S2N_ERROR_IF(op->complete, S2N_ERR_ASYNC_ALREADY_PERFORMED);
+    ENSURE_POSIX_NONNULL(op);
+    ENSURE_POSIX(!op->complete, S2N_ERR_ASYNC_ALREADY_PERFORMED);
 
-    const struct s2n_async_pkey_op_actions *actions = s2n_async_get_actions(op->type);
-    notnull_check(actions);
+    const struct s2n_async_pkey_op_actions *actions = NULL;
+    GUARD_AS_POSIX(s2n_async_get_actions(op->type, &actions));
 
-    GUARD(actions->perform(op, key));
+    GUARD_AS_POSIX(actions->perform(op, key));
 
     op->complete = 1;
 
-    return 0;
+    return S2N_SUCCESS;
 }
 
 int s2n_async_pkey_op_apply(struct s2n_async_pkey_op *op, struct s2n_connection *conn)
 {
-    notnull_check(op);
-    S2N_ERROR_IF(!op->complete, S2N_ERR_ASYNC_NOT_PERFORMED);
-    S2N_ERROR_IF(op->applied, S2N_ERR_ASYNC_ALREADY_APPLIED);
+    ENSURE_POSIX_NONNULL(op);
+    ENSURE_POSIX(op->complete, S2N_ERR_ASYNC_NOT_PERFORMED);
+    ENSURE_POSIX(!op->applied, S2N_ERR_ASYNC_ALREADY_APPLIED);
     /* We could have just used op->conn and removed a conn argument, but we want caller
      * to be explicit about connection it wants to resume. Plus this gives more
      * protections in cases if caller frees connection object and then tries to resume
      * the connection. */
-    S2N_ERROR_IF(op->conn != conn, S2N_ERR_ASYNC_WRONG_CONNECTION);
-    S2N_ERROR_IF(conn->handshake.async_state == S2N_ASYNC_INVOKING_CALLBACK, S2N_ERR_ASYNC_APPLY_WHILE_INVOKING);
-    S2N_ERROR_IF(conn->handshake.async_state != S2N_ASYNC_INVOKED_WAITING, S2N_ERR_ASYNC_WRONG_CONNECTION);
+    ENSURE_POSIX(op->conn == conn, S2N_ERR_ASYNC_WRONG_CONNECTION);
+    ENSURE_POSIX(conn->handshake.async_state != S2N_ASYNC_INVOKING_CALLBACK, S2N_ERR_ASYNC_APPLY_WHILE_INVOKING);
+    ENSURE_POSIX(conn->handshake.async_state == S2N_ASYNC_INVOKED_WAITING, S2N_ERR_ASYNC_WRONG_CONNECTION);
```

Comment:
Is `S2N_ERR_ASYNC_WRONG_CONNECTION` the correct error if the state isn't `S2N_ASYNC_INVOKED_WAITING`?

Reply:
The logic here is that we passed `op.conn == conn` check so the struct is the same, but if it's in a wrong state that likely means that connection was wiped and reused for something else, so we use `S2N_ERR_ASYNC_WRONG_CONNECTION`.